### ISSUE TITLE
Enhance model

### ISSUE THREAD

**Issue by alo8** [neherlab/covid19_scenarios#116]
Proposal to improve the model used to provide more realistic results.

Current set of differential equations (just using the infected individuals as an example):
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\mathrm{d}&space;I(t)&space;}{\mathrm{d}&space;t}=&space;E(t)/t_l-I(t)/t_i" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\mathrm{d}&space;I(t)&space;}{\mathrm{d}&space;t}=&space;E(t)/t_l-I(t)/t_i" title="\frac{\mathrm{d} I(t) }{\mathrm{d} t}= E(t)/t_l-I(t)/t_i" /></a>

Proposed set of differential equations (the other functions accordingly):
<a href="https://www.codecogs.com/eqnedit.php?latex=\frac{\mathrm{d}&space;I(t)&space;}{\mathrm{d}&space;t}=&space;\frac{\mathrm{d}&space;E(t-t_l)&space;}{\mathrm{d}&space;t}-\frac{\mathrm{d}&space;I(t-t_i)&space;}{\mathrm{d}&space;t}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\frac{\mathrm{d}&space;I(t)&space;}{\mathrm{d}&space;t}=&space;\frac{\mathrm{d}&space;E(t-t_l)&space;}{\mathrm{d}&space;t}-\frac{\mathrm{d}&space;I(t-t_i)&space;}{\mathrm{d}&space;t}" title="\frac{\mathrm{d} I(t) }{\mathrm{d} t}= \frac{\mathrm{d} E(t-t_l) }{\mathrm{d} t}-\frac{\mathrm{d} I(t-t_i) }{\mathrm{d} t}" /></a>

Why is that set of equations better? Because the number of newly infectious individuals does not depend on the number of the currently exposed individuals but rather on the number of newly exposed individuals <a href="https://www.codecogs.com/eqnedit.php?latex=t_l" target="_blank"><img src="https://latex.codecogs.com/gif.latex?t_l" title="t_l" /></a> days ago. The same logic can be applied to the other differential equations.

This new set of differential equations can be solved in the same iterative way as the current model. Only slight modifications are necessary. E.g.
Currently:    newInfectious = exposed / incubationTime
Proposed:  newInfectious = newCases[-incubationTime]

With this proposed model you can especially get a more realistic time delay between the number of confirmed cases and the number of dead individuals. This can be verified/fitted to existing data.

**Comment by rneher** [comment]
Thanks @alo8. Turning this into a delay differential equation might be useful. We could eliminate the exposed state in that case. but specifying the initial condition requires more care in that case. But having the exposed state already introduces a delay, so we don't expect this to change the model qualitatively.

**Comment by alo8** [comment]
I made a simple, quick comparison of the current model and this proposal: (No averaging/filtering applied.  Sorry for so many lines in one graph)
![pic1](https://user-images.githubusercontent.com/8531651/77258697-9df44f80-6c7c-11ea-8978-ec1414ba9a48.jpg)
Scenario parameters are as shown here:
![pic2](https://user-images.githubusercontent.com/8531651/77258699-9e8ce600-6c7c-11ea-8bda-f61f65ab56e0.jpg)

**Comment by nnoll** [comment]
@alo8: thanks for your suggestions! I have a few clarifying questions.

Are you proposing to remove the exposed class and replace with a time-delay in it's place as your proposal of newInfectious = newCases[-latencyPeriod] suggests? Or do you want to add an overall time-delay between the number of infected vs exposed and leave the subpopulations the same?

If it's the former, then I'm not sure I expect huge qualitative differences between the current version and the proposal - in effect we are computing the mean/expected number of infections given an exponential distribution of latency times vs the proposal as a well-defined singular time of latency.

If it's the later, after working through some of the math, I think the qualitatively different effect is that the time delay is expected to add oscillatory dynamics (you see this in your simulation curves) and potentially (not sure here) a Hopf bifurcation at large enough time delays (not relevant for our present goals).  I also think care would also have to be taken to ensure sub-populations don't become negative.

Let me know if I misunderstood something.

**Comment by rneher** [comment]
One other aspect in your simulation output I am puzzled by is the rapid decrease of the critically ill after the peak. Did you replace all equations by delay equations?

**Comment by alo8** [comment]
@nnol, @rneher: thanks for your interest and feedback!

Let me try to explain the proposal in more detail. 
I refer to the nomenclature used in model.js and run.js:

If _newPop_ would also store the values of _newCases_,_newInfectious_,... for each time step, then one could access in each time step iteration not only the last elements (_pop_) but also older ones from the _dynamics_ list. For simplicity here let us denote newCases[n] as the new cases at time iteration n. (not considering age subpopulations here). Then the proposed new model looks like this:

newCases[n] = infectionRate infectious[n-1] (susceptible[n-1]/population);  
newInfectious[n] = newCases[n-incubationPeriod];
newRecovered[n] =  newInfectious[n-infectiousPeriod] recoveryRate;
newHospitalized[n] = newInfectious[n-infectiousPeriod] hospitalizedRate;
newDischarged[n] = newHospitalized[n-hospitalPeriod] dischargeRate;
newCritical[n] = newHospitalized[n-hospitalPeriod] criticalRate;
newStabilized[n] = newCritical[n-ICUStayPeriod] stabilizationRate;
newICUDead[n] = newCritical[n-ICUStayPeriod] fatalRate;

So yes, all equations here consider the respective delay.
The calculation of _susceptibe_, _exposed_, _infectious_,... can stay unchanged.

For analytical purposes, the full set of differential equations can be derived from these delta definitions.

If you start with one infectious person (_numCases_=1) the initial conditions are simply all zero for n<=0, and newInfectious[1]=1. The total number of individuals stays the same and no negative delta values occure.

Yes, some oscillatory dynamics can be seen at the beginning of the simulation. This could be prevented by adding a slight variation of the delay times according to some random distribution. These random distibution functions could bring the model even closer to reality.
A bifurcation I did not encounter during the simulations so far.

In general the current model leads to significantly shorter delays and slightly faster growth during the exponential growth phase. For a linear growth the results of current and proposed model would be equal.

Yes, the difference in the decrease of the critically is significant. I did not think about why yet...
